internal/controller/dataplane/controller_finalize.go

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2025, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package dataplane
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	choreov1 "github.com/openchoreo/openchoreo/api/v1"
+	"github.com/openchoreo/openchoreo/internal/controller"
+	"github.com/openchoreo/openchoreo/internal/labels"
+)
+
+// DataPlaneCleanupFinalizer is the finalizer that is used to clean up dataplane resources.
+const DataPlaneCleanupFinalizer = "core.choreo.dev/dataplane-cleanup"
+
+// ensureFinalizer ensures that the finalizer is added to the dataplane.
+// The first return value indicates whether the finalizer was added to the dataplane.
+func (r *Reconciler) ensureFinalizer(ctx context.Context, dataPlane *choreov1.DataPlane) (bool, error) {
+	// If the dataplane is being deleted, no need to add the finalizer
+	if !dataPlane.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	if controllerutil.AddFinalizer(dataPlane, DataPlaneCleanupFinalizer) {
+		return true, r.Update(ctx, dataPlane)
+	}
+
+	return false, nil
+}
+
+func (r *Reconciler) finalize(ctx context.Context, old, dataPlane *choreov1.DataPlane) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("dataplane", dataPlane.Name)
+
+	if !controllerutil.ContainsFinalizer(dataPlane, DataPlaneCleanupFinalizer) {
+		return ctrl.Result{}, nil
+	}
+
+	// Mark the condition as finalizing and return so that the dataplane will indicate that it is being finalized.
+	// The actual finalization will be done in the next reconcile loop triggered by the status update.
+	if meta.SetStatusCondition(&dataPlane.Status.Conditions, NewDataPlaneFinalizingCondition(dataPlane.Generation)) {
+		if err := controller.UpdateStatusConditions(ctx, r.Client, old, dataPlane); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Perform cleanup logic for referenced environments
+	artifactsDeleted, err := r.deleteEnvironmnetsAndWait(ctx, dataPlane)
+	if err != nil {
+		logger.Error(err, "Failed to delete environments")
+		return ctrl.Result{}, err
+	}
+	if !artifactsDeleted {
```

Comment:
```suggestion
	// Perform cleanup logic for referenced environments
	envsDeleted, err := r.deleteEnvironmnetsAndWait(ctx, dataPlane)
	if err != nil {
		logger.Error(err, "Failed to delete environments")
		return ctrl.Result{}, err
	}
	if !envsDeleted {
```

Can we consider an environment to be an artifact?